rca/shortcourses/access_planit.py

```diff
@@ -0,0 +1,151 @@
+import logging
+
+import requests
+from bs4 import BeautifulSoup as bs
+from django.conf import settings
+from django.core.cache import cache
+from django.http.request import QueryDict
+from django.utils.dateparse import parse_datetime
+from requests.exceptions import Timeout
+
+
+"""
+Provides functionality for fetching data from access planit xml feed
+"""
+
+
+# TODO remove this, it's just for debugging
+def print_message(msg):
+    print("\n")
+    print("=" * len(msg))
+    print(f"{msg}")
+    print("=" * len(msg))
+    print("\n")
+
+
+logger = logging.getLogger(__name__)
+
+
+class AccessPlanitException(Exception):
+    pass
+
+
+class AccessPlanitXMLParser:
+    """ XML parses
+    xml: a blob of xml to a list (from response.text)
+    """
+
+    def __init__(self, xml):
+        self.xml = xml
+
+    def parse(self):
+        """ Parses the given xml to a list
+            return: list of dicts, or an empty list
+        """
+        bs_content = bs(self.xml, "lxml")
+        course_dates = bs_content.find("dates").find_all("wicoursedate")
+        items = []
+        if course_dates:
+            for i in course_dates:
+                item = {
+                    "course_date_id": i.coursedateid.text,
+                    "course_id": i.courseid.text,
+                    "cost": i.cost.text,
+                    "start_date": parse_datetime(i.startdate.text),
+                    "end_date": parse_datetime(i.enddate.text),
+                    "spaces_available": i.spacesavailable.text,
+                    "status": i.status.text,
+                    "book_now_url": i.booknowurl.text,
+                    "enquire_url": i.enquireurl.text,
+                }
+                items.append(item)
+
+        return items
+
+    def get_parsed_data(self):
+        """
+        Returns:
+            list -- A list containing parsed data
+        """
+        parsed_data = self.parse()
+        return parsed_data
+
+
+class AccessPlanitXML:
+    """Used for retrieving short course data from the AccessPlanit xml feeds.
+    The data is periodically fetched with a management command and set
+    in the low-level cache.
+
+    Raises:
+        AccessPlanitException: Should the fetch encouter Timeouts or any other
+        exceptions, this raised and caught when setting the data in the cache.
+    """
+
+    def __init__(self, course_id):
+        # This might be better as a default on the model
+        # but if there is no course ID passed we seem to get some default xml
+        # data back, so set it as 0
+        self.course_id = course_id if course_id else "0"
+        self.company_id = "ROYALC9RCH"  # TODO - settings field perhaps?
```

Comment:
Perhaps make this a settings field if time allows, as it might be different in the live instance of the XML, at the moment we are using a sandbox

Reply:
Thanks, Done